File: MathematicFunctions/MathematicFunctions.cs
namespace Molytho.TowerOfHanoi
{
    /// <summary>
    /// Provides static functions for the Tower-of-Hanoi-MoveCount-problem
    /// </summary>
    public static class MathematicFunctions
    {
        /// <summary>
        /// Calculates the Increment of a <paramref name="pegCount"/>, <paramref name="diskCount"/> combination
        /// </summary>
        /// <param name="pegCount">Count of pegs</param>
        /// <param name="diskCount">Count of disks</param>
        /// <returns>Increment of the combination</returns>
        public static ulong Increment(ulong pegCount, ulong diskCount)
        {
            if(pegCount < 3)
                throw new System.ArgumentException("pegCounts smaller than 3 are senseless", "pegCount");
            if(diskCount == 0)
                return 0;
            ulong increment = 0, temp = 1;

            while(diskCount > temp)
            {
                increment++;
                temp = temp * (pegCount - 2 + increment) / increment;
            }

            return increment + 1;
        }
        /// <summary>
        /// Calculates minimum moves needed for solving the Tower-of-Hanoi with 3+ pegs
        /// </summary>
        /// <param name="increment">The increment of the <paramref name="pegCount"/>, <paramref name="diskCount"/> combination</param>
        /// <param name="pegCount">The peg count of the Tower-of-Hanoi game</param>
        /// <param name="diskCount">The disk count of the Tower-of-Hanoi game</param>
        /// <returns>The move count minimaly needed</returns>
        private static ulong MoveCountIntern(ulong increment, ulong pegCount, ulong diskCount)
        {
            ulong temp = diskCount * HelperFunctions.Pow(2, increment - 1);

            for(ulong i = 0; i < increment; i++)
            {
                temp += HelperFunctions.Pow(2, i) * HelperFunctions.BinomialCoefficent(i + pegCount - 3, pegCount - 3);
            }
            for(ulong i = 0; i < increment; i++)
            {
                temp -= HelperFunctions.Pow(2, increment - 1) * HelperFunctions.BinomialCoefficent(i + pegCount - 3, pegCount - 3);
            }

            return temp;
        } //Implementation for the formular

        /// <summary>
        /// Calculates minimum moves needed for solving the Tower-of-Hanoi with 3+ pegs
        /// </summary>
        /// <param name="pegCount">The peg count of the Tower-of-Hanoi game</param>
        /// <param name="diskCount">The disk count of the Tower-of-Hanoi game</param>
        /// <returns>The move count minimaly needed</returns>
        public static ulong MoveCount(ulong pegCount, ulong diskCount)
        {
            if(pegCount < 3) //Less than 3 pegs is nonsense
                throw new System.ArgumentException("pegCounts smaller than 3 are senseless", "pegCount");
            if(pegCount == 3) //3 pegs is simply 2^diskCount - 1
            {
                return HelperFunctions.Pow(2, diskCount) - 1;
            }

            ulong increment;

            increment = Increment(pegCount, diskCount); //Calculate the increment
            return MoveCountIntern(increment, pegCount, diskCount); //and pass it to the intern function
        }
    }
}

File: cpp/MathematicFunctions/MathematicFunctions.cpp
#include "MathematicFunctions.h"

#include <cstdint>

static uint64_t Pow(uint64_t base, uint64_t exponent)
{
	uint64_t temp = 1;
	for (uint64_t i = 1; i <= exponent; i++)
		temp *= base;
	return temp;
}

static uint64_t CalculateMoveCountIntern(const uint64_t increment, const uint64_t pegCount, const uint64_t diskCount)
{
	uint64_t temp = diskCount * Pow(2, increment - 1);

	for (uint64_t i = 1; i < increment + 1; i++)
	{
		temp += Pow(2, i - 1) * BinomialCoefficent(i + pegCount - 4, pegCount - 3);
	}
	for (uint64_t i = 1; i < increment + 1; i++)
	{
		temp -= Pow(2, increment - 1) * BinomialCoefficent(i + pegCount - 4, pegCount - 3);
	}

	return temp;
}

uint64_t BinomialCoefficent(uint64_t n, uint64_t k)
{
	if (2 * k > n) k = n - k;
	if (k == 0) return 1;
	if (k < 1) return 0;
	uint64_t ergebnis = 1;
	for (uint64_t i = 1; i <= k; i++)
	{
		ergebnis = ergebnis * (n - k + i) / i;
	}
	return ergebnis;
}

uint64_t CalculateIncrement(const uint64_t pegCount, const uint64_t diskCount)
{
	if (diskCount == 0) return 0;
	uint64_t increment = 0;
	uint64_t temp = 1;

	while (diskCount > temp)
	{
		increment++;
		temp = BinomialCoefficent(pegCount - 2 + increment, pegCount - 2);
	}

	return increment + 1;
}

uint64_t CalculateMoveCount(const uint64_t pegCount, const uint64_t diskCount)
{
	uint64_t ret;
	if (pegCount == 3)
	{
		ret = Pow(2, diskCount) - 1;
		return ret;
	}

	uint64_t increment;

	increment = CalculateIncrement(pegCount, diskCount);
	ret = CalculateMoveCountIntern(increment, pegCount, diskCount);

	return ret;
}

